File: demos/54-orderpipe/fulfillment/events.py
"""Forge Events publish/consume client for OrderPipe fulfillment (epic 54.04)."""

from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any


@dataclass
class EventsConfig:
    base_url: str
    source: str
    consumer: str
    identity: str
    subject: str
    publish_subject: str
    ack_wait_s: int
    max_deliveries: int
    poll_ms: int
    batch: int


def _env_int(env: dict[str, str], key: str, default: int) -> int:
    raw = (env.get(key) or "").strip()
    if not raw:
        return default
    try:
        n = int(raw)
    except ValueError:
        return default
    return n if n > 0 else default


def load_events_config(environ: dict[str, str] | None = None) -> EventsConfig:
    env = environ if environ is not None else os.environ
    base = (env.get("FORGE_EVENTS_URL") or "").strip() or "http://host.docker.internal:4105"
    source = (env.get("FORGE_SERVICE_NAME") or "").strip() or "orderpipe-fulfillment"
    consumer = (env.get("FORGE_EVENTS_CONSUMER") or "").strip() or "orderpipe-fulfill"
    identity = (env.get("FORGE_EVENTS_CONSUMER_IDENTITY") or "").strip() or consumer
    subject = (env.get("FORGE_EVENTS_SUBJECT") or "").strip() or "order.charged"
    publish_subject = (env.get("FORGE_EVENTS_PUBLISH_SUBJECT") or "").strip() or "order.fulfilled"
    return EventsConfig(
        base_url=base.rstrip("/"),
        source=source,
        consumer=consumer,
        identity=identity,
        subject=subject,
        publish_subject=publish_subject,
        ack_wait_s=_env_int(env, "FORGE_DEFAULT_ACK_WAIT_S", 30),
        max_deliveries=_env_int(env, "FORGE_DEFAULT_MAX_DELIVERIES", 5),
        poll_ms=_env_int(env, "FORGE_EVENTS_POLL_MS", 400),
        batch=_env_int(env, "FORGE_EVENTS_BATCH", 8),
    )


@dataclass
class DeliveredMessage:
    event_id: str
    subject: str
    ack_token: str
    delivery_count: int
    data: dict[str, Any]


class EventsClient:
    def __init__(self, cfg: EventsConfig | None = None) -> None:
        self.cfg = cfg or load_events_config()

    def _request(
        self,
        method: str,
        url: str,
        data: bytes | None = None,
        headers: dict[str, str] | None = None,
        timeout: float = 45.0,
    ) -> tuple[int, bytes]:
        req = urllib.request.Request(url, data=data, method=method, headers=headers or {})
        try:
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                return int(resp.status), resp.read()
        except urllib.error.HTTPError as exc:
            body = exc.read() if exc.fp else b""
            return int(exc.code), body

    def ping(self) -> None:
        code, body = self._request("GET", f"{self.cfg.base_url}/health/ready", timeout=10.0)
        if code != 200:
            raise RuntimeError(f"events ready HTTP {code}: {body[:256]!r}")

    def ensure_consumer(self) -> None:
        payload = {
            "name": self.cfg.consumer,
            "subject": self.cfg.subject,
            "ack_wait_s": self.cfg.ack_wait_s,
            "max_deliveries": self.cfg.max_deliveries,
            "identity": self.cfg.identity,
        }
        raw = json.dumps(payload).encode()
        code, body = self._request(
            "POST",
            f"{self.cfg.base_url}/v1/consumers",
            data=raw,
            headers={"Content-Type": "application/json"},
            timeout=15.0,
        )
        if code not in (200, 201):
            raise RuntimeError(f"create consumer HTTP {code}: {body[:512]!r}")

    def consume(self) -> list[DeliveredMessage]:
        payload = {"consumer": self.cfg.consumer, "batch": self.cfg.batch}
        raw = json.dumps(payload).encode()
        code, body = self._request(
            "POST",
            f"{self.cfg.base_url}/v1/consume",
            data=raw,
            headers={"Content-Type": "application/json"},
            timeout=45.0,
        )
        if code != 200:
            raise RuntimeError(f"consume HTTP {code}: {body[:512]!r}")
        parsed = json.loads(body.decode() or "{}")
        out: list[DeliveredMessage] = []
        for item in parsed.get("messages") or []:
            data = item.get("data") or {}
            if isinstance(data, str):
                try:
                    data = json.loads(data)
                except json.JSONDecodeError:
                    data = {}
            if not isinstance(data, dict):
                data = {}
            out.append(
                DeliveredMessage(
                    event_id=str(item.get("event_id") or ""),
                    subject=str(item.get("subject") or ""),
                    ack_token=str(item.get("ack_token") or ""),
                    delivery_count=int(item.get("delivery_count") or 0),
                    data=data,
                )
            )
        return out

    def mark_processed(self, event_id: str) -> None:
        raw = json.dumps({"consumer": self.cfg.consumer, "event_id": event_id}).encode()
        code, body = self._request(
            "POST",
            f"{self.cfg.base_url}/v1/processed",
            data=raw,
            headers={"Content-Type": "application/json"},
            timeout=15.0,
        )
        if code not in (200, 204):
            raise RuntimeError(f"processed HTTP {code}: {body[:256]!r}")

    def ack(self, ack_token: str) -> None:
        raw = json.dumps({"ack_token": ack_token}).encode()
        code, body = self._request(
            "POST",
            f"{self.cfg.base_url}/v1/ack",
            data=raw,
            headers={"Content-Type": "application/json"},
            timeout=15.0,
        )
        if code not in (200, 204):
            raise RuntimeError(f"ack HTTP {code}: {body[:256]!r}")

    def nak(self, ack_token: str, delay_s: int = 0) -> None:
        payload: dict[str, Any] = {"ack_token": ack_token}
        if delay_s > 0:
            payload["delay_s"] = delay_s
        raw = json.dumps(payload).encode()
        code, body = self._request(
            "POST",
            f"{self.cfg.base_url}/v1/nak",
            data=raw,
            headers={"Content-Type": "application/json"},
            timeout=15.0,
        )
        if code not in (200, 204):
            raise RuntimeError(f"nak HTTP {code}: {body[:256]!r}")

    def publish_order_fulfilled(
        self,
        order_id: str,
        customer_email: str,
        total_cents: int,
        occurred_at: str | None = None,
    ) -> None:
        occurred_at = occurred_at or datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
        payload = {
            "subject": self.cfg.publish_subject,
            "source": self.cfg.source,
            "data": {
                "order_id": order_id,
                "customer_email": customer_email,
                "status": "fulfilled",
                "total_cents": int(total_cents),
                "occurred_at": occurred_at,
                "source": self.cfg.source,
            },
        }
        raw = json.dumps(payload).encode()
        headers = {
            "Content-Type": "application/json",
            "Idempotency-Key": f"{order_id}:{self.cfg.publish_subject}",
        }
        code, body = self._request(
            "POST",
            f"{self.cfg.base_url}/v1/events",
            data=raw,
            headers=headers,
            timeout=15.0,
        )
        if code != 202:
            raise RuntimeError(f"events publish HTTP {code}: {body[:1024]!r}")

File: demos/54-orderpipe/api/events.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"
)

const (
	subjectPlaced    = "order.placed"
	subjectValidated = "order.validated"
	subjectCharged   = "order.charged"
	subjectFulfilled = "order.fulfilled"
	subjectNotified  = "order.notified"
)

type eventsConfig struct {
	BaseURL       string
	Source        string
	AckWaitS      int
	MaxDeliveries int
	PollMS        int
	Batch         int
	// Consumer names (one durable per subject stage).
	ValidateConsumer  string
	ChargeConsumer    string
	FulfilledConsumer string
	NotifiedConsumer  string
}

func loadEventsConfig() eventsConfig {
	base := strings.TrimSpace(os.Getenv("FORGE_EVENTS_URL"))
	if base == "" {
		base = "http://host.docker.internal:4105"
	}
	source := strings.TrimSpace(os.Getenv("FORGE_SERVICE_NAME"))
	if source == "" {
		source = "orderpipe-api"
	}
	return eventsConfig{
		BaseURL:           strings.TrimRight(base, "/"),
		Source:            source,
		AckWaitS:          envInt("FORGE_DEFAULT_ACK_WAIT_S", 30),
		MaxDeliveries:     envInt("FORGE_DEFAULT_MAX_DELIVERIES", 5),
		PollMS:            envInt("FORGE_EVENTS_POLL_MS", 400),
		Batch:             envInt("FORGE_EVENTS_BATCH", 8),
		ValidateConsumer:  envOr("FORGE_EVENTS_CONSUMER_VALIDATE", "orderpipe-validate"),
		ChargeConsumer:    envOr("FORGE_EVENTS_CONSUMER_CHARGE", "orderpipe-charge"),
		FulfilledConsumer: envOr("FORGE_EVENTS_CONSUMER_FULFILLED", "orderpipe-mark-fulfilled"),
		NotifiedConsumer:  envOr("FORGE_EVENTS_CONSUMER_NOTIFIED", "orderpipe-mark-notified"),
	}
}

func envInt(key string, def int) int {
	raw := strings.TrimSpace(os.Getenv(key))
	if raw == "" {
		return def
	}
	n, err := strconv.Atoi(raw)
	if err != nil || n <= 0 {
		return def
	}
	return n
}

type eventsClient struct {
	cfg        eventsConfig
	httpClient *http.Client
}

func newEventsClient(cfg eventsConfig) *eventsClient {
	return &eventsClient{
		cfg: cfg,
		httpClient: &http.Client{
			Timeout: 45 * time.Second,
		},
	}
}

func (c *eventsClient) enabled() bool {
	return c != nil && strings.TrimSpace(c.cfg.BaseURL) != ""
}

func (c *eventsClient) Ping(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, c.cfg.BaseURL+"/health/ready", nil)
	if err != nil {
		return err
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 256))
		return fmt.Errorf("events ready HTTP %d: %s", resp.StatusCode, string(b))
	}
	return nil
}

type orderEventData struct {
	OrderID       string `json:"order_id"`
	CustomerEmail string `json:"customer_email"`
	Status        string `json:"status"`
	TotalCents    int    `json:"total_cents"`
	OccurredAt    string `json:"occurred_at"`
	Source        string `json:"source,omitempty"`
}

func (c *eventsClient) PublishOrderEvent(ctx context.Context, subject string, order *Order) error {
	if !c.enabled() {
		return fmt.Errorf("events url not configured")
	}
	if order == nil {
		return fmt.Errorf("order is required")
	}
	status := strings.TrimPrefix(subject, "order.")
	body := map[string]any{
		"subject": subject,
		"source":  c.cfg.Source,
		"data": orderEventData{
			OrderID:       order.ID,
			CustomerEmail: order.CustomerEmail,
			Status:        status,
			TotalCents:    order.TotalCents,
			OccurredAt:    time.Now().UTC().Format(time.RFC3339),
			Source:        c.cfg.Source,
		},
	}
	raw, err := json.Marshal(body)
	if err != nil {
		return err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.cfg.BaseURL+"/v1/events", bytes.NewReader(raw))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Idempotency-Key", order.ID+":"+subject)
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if resp.StatusCode != http.StatusAccepted {
		return fmt.Errorf("events publish HTTP %d: %s", resp.StatusCode, string(respBody))
	}
	return nil
}

type consumerBinding struct {
	Name    string
	Subject string
}

func (c *eventsClient) bindings() []consumerBinding {
	// validate/charge are owned by the order-saga driver (54.05); API only
	// mirrors fulfill/notify completions onto order status + saga_events.
	return []consumerBinding{
		{Name: c.cfg.FulfilledConsumer, Subject: subjectFulfilled},
		{Name: c.cfg.NotifiedConsumer, Subject: subjectNotified},
	}
}

func (c *eventsClient) EnsureConsumers(ctx context.Context) error {
	for _, b := range c.bindings() {
		if err := c.ensureConsumer(ctx, b); err != nil {
			return err
		}
	}
	return nil
}

func (c *eventsClient) ensureConsumer(ctx context.Context, b consumerBinding) error {
	body := map[string]any{
		"name":           b.Name,
		"subject":        b.Subject,
		"ack_wait_s":     c.cfg.AckWaitS,
		"max_deliveries": c.cfg.MaxDeliveries,
		"identity":       b.Name,
	}
	raw, _ := json.Marshal(body)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.cfg.BaseURL+"/v1/consumers", bytes.NewReader(raw))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusCreated || resp.StatusCode == http.StatusOK {
		return nil
	}
	bbody, _ := io.ReadAll(io.LimitReader(resp.Body, 512))
	return fmt.Errorf("create consumer %s HTTP %d: %s", b.Name, resp.StatusCode, string(bbody))
}

type deliveredMessage struct {
	EventID       string          `json:"event_id"`
	Subject       string          `json:"subject"`
	AckToken      string          `json:"ack_token"`
	DeliveryCount int             `json:"delivery_count"`
	Data          json.RawMessage `json:"data"`
}

type consumeResponse struct {
	Messages []deliveredMessage `json:"messages"`
}

func (c *eventsClient) Consume(ctx context.Context, consumer string) ([]deliveredMessage, error) {
	body := map[string]any{
		"consumer": consumer,
		"batch":    c.cfg.Batch,
	}
	raw, _ := json.Marshal(body)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.cfg.BaseURL+"/v1/consume", bytes.NewReader(raw))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	respBody, err := io.ReadAll(io.LimitReader(resp.Body, 4<<20))
	if err != nil {
		return nil, err
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("consume HTTP %d: %s", resp.StatusCode, string(respBody))
	}
	var out consumeResponse
	if err := json.Unmarshal(respBody, &out); err != nil {
		return nil, err
	}
	return out.Messages, nil
}

func (c *eventsClient) MarkProcessed(ctx context.Context, consumer, eventID string) error {
	body := map[string]any{"consumer": consumer, "event_id": eventID}
	raw, _ := json.Marshal(body)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.cfg.BaseURL+"/v1/processed", bytes.NewReader(raw))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 256))
		return fmt.Errorf("processed HTTP %d: %s", resp.StatusCode, string(b))
	}
	return nil
}

func (c *eventsClient) Ack(ctx context.Context, ackToken string) error {
	body := map[string]any{"ack_token": ackToken}
	raw, _ := json.Marshal(body)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.cfg.BaseURL+"/v1/ack", bytes.NewReader(raw))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 256))
		return fmt.Errorf("ack HTTP %d: %s", resp.StatusCode, string(b))
	}
	return nil
}

func (c *eventsClient) Nak(ctx context.Context, ackToken string, delayS int) error {
	body := map[string]any{"ack_token": ackToken}
	if delayS > 0 {
		body["delay_s"] = delayS
	}
	raw, _ := json.Marshal(body)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.cfg.BaseURL+"/v1/nak", bytes.NewReader(raw))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 256))
		return fmt.Errorf("nak HTTP %d: %s", resp.StatusCode, string(b))
	}
	return nil
}
